pkg/interactor/save_data_test.go

```diff
@@ -0,0 +1,81 @@
+package interactor
+
+import (
+	"errors"
+	"testing"
+	"time"
+
+	"github.com/CESARBR/knot-cloud-storage/pkg/entities"
+	"github.com/CESARBR/knot-cloud-storage/pkg/mocks"
+	"github.com/stretchr/testify/assert"
+)
+
+var date = time.Date(2020, time.April, 15, 0, 0, 0, 0, time.UTC)
+
+type SaveDataTestCase struct {
+	name             string
+	authorization    string
+	deviceID         string
+	payload          []entities.Payload
+	data             entities.Data
+	fakeLogger       *mocks.FakeLogger
+	fakeDataStore    *mocks.FakeDataStore
+	fakeThingService *mocks.FakeThingService
+}
+
+var sdCases = []SaveDataTestCase{
+	{
+		"authorization key not provided",
+		"",
+		"",
+		nil,
+		entities.Data{},
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeThingService{ReturnErr: errors.New("no access token provided")},
+	},
+	{
+		"data successfully saved on the database",
+		"authorization-key",
+		"fc3fcf912d0c290a",
+		[]entities.Payload{{SensorID: 1, Value: 999}},
+		entities.Data{From: "fc3fcf912d0c290a", Payload: entities.Payload{SensorID: 1, Value: 999}, Timestamp: date},
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeThingService{},
+	},
+	{
+		"failed to save the data on the database",
+		"authorization-key",
+		"fc3fcf912d0c290a",
+		[]entities.Payload{{SensorID: 1, Value: 999}},
+		entities.Data{From: "fc3fcf912d0c290a", Payload: entities.Payload{SensorID: 1, Value: 999}, Timestamp: date},
+		&mocks.FakeLogger{},
+		&mocks.FakeDataStore{},
+		&mocks.FakeThingService{},
+	},
+}
+
+func TestSaveData(t *testing.T) {
+	for _, tc := range sdCases {
+		t.Run(tc.name, func(t *testing.T) {
+			tc.fakeDataStore.
+				On("Save", tc.data).
+				Return(tc.fakeDataStore.ReturnErr).
+				Maybe()
+			tc.fakeThingService.
+				On("List", tc.authorization).
+				Return(tc.fakeThingService.Thing, tc.fakeThingService.ReturnErr).
+				Maybe()
+		})
+
+		dataInteractor := NewDataInteractor(tc.fakeThingService, tc.fakeDataStore, tc.fakeLogger)
+		err := dataInteractor.Save(tc.authorization, tc.deviceID, tc.payload, date)
+		if tc.authorization == "" {
+			assert.EqualError(t, err, "no access token provided")
+		}
```

Comment:
Suggestion for when you update the test case structure, as @joaoaneto commented [here](https://github.com/CESARBR/knot-cloud-storage/pull/29/commits/5eebce7fc9fa7d6be99a5552e4ebe7de913e3054#r417637923)

```suggestion
		if tc.authorization == "" {
			msg := tc.expectedErr.Error()
			assert.EqualError(t, err, msg)
		}
```

Reply:
Done!